### ISSUE TITLE
DashboardLayout: Match multiple routes with a single navigation item

### ISSUE THREAD

**Issue by Janpot** [mui/mui-toolpad#3974]
### Search keywords

multiple routes

### Latest version

- [X] I have tested the latest version

### Summary

I have a routing structure that has dynamic segments such as `/orders/:id`. I want a single navigation item in the sidebar that can represent all these routes.

### option 1

Expand the `segment` parameter to include multiple route segments:

```tsx
{
  segment: 'orders/:id',
  title: 'Orders'
}
```
Question: what should this item link to then when clicked?

### option 2

Add another property to expand what this route can match on

```tsx
{
  segment: 'orders',
  alias: 'orders/:id'
  title: 'Orders'
}
```

### option 3

Add new item type that allows defining segments that alias to the parent

```tsx
{
  segment: 'orders',
  title: 'Orders',
  children: [
    {
      kind: 'alias',
      segment: ':id'
    }
  ]
}
```

### option 4

?

## benchmark

To Do

## Notes

* With whatever we choose, how should page container behave?
* Should it Probably also support deeply nested aliases, e.g. `/users/:userId/invoices/:invoiceId`
* Currently we warn when a router path is matching multiple navigation items. Should we instead start thinking about a fallback behavior?
    e.g. considering the structure:
    - `/orders` => orders overview page
    - `/orders/new` => new order form (even though it also matches `/orders/:id`)
    - `/orders/:id` => fallback to
* As the behavior gets more complex. We should spec it out and document it well.

### Motivation

https://discord.com/channels/1131323012554174485/1248515176626524200/1275616786800181249

**Comment by apedroferreira** [comment]
I see...
I think option 2 sounds best as even from the user side it would not create confusion about the default page. 
It's also probably more flexible. Other possible names for `pattern` could be maybe `match` for example, but `pattern` sounds good.

Option 1 seems a bit worse to me  from that point of view, and option 3 I would think is the worst as some children wouldn't be visual items and that just adds even more possible confusion.

> With whatever we choose, how should page container behave?

We should probably make it work automatically with that kind of parameters and add an extra "segment" to the breadcrumb.

> Should it Probably also support deeply nested aliases, e.g. /users/:userId/invoices/:invoiceId

I would say yes.

> Currently we warn when a router path is matching multiple navigation items. Should we instead start thinking about a fallback behavior?

Seems like a complicated problem, not sure if it would be too much to ask users to exclude "new" from the regex if they really want to do something like this... Or we could give priority to absolute paths? Just some first thoughts.

> If we go with option 2, should it also include absolute paths?

 I guess any regex could be used to match whatever people want? But we could accept a specific string too maybe I guess.

**Comment by Janpot** [comment]
> I think option 2 sounds best as even from the user side it would not create confusion about the default page.
It's also probably more flexible. Other possible names for pattern could be maybe match for example, but pattern sounds good.

Yes, 2. makes most sense. I chose the "pattern" nomenclature to align with https://developer.mozilla.org/en-US/docs/Web/API/URLPattern

> We should probably make it work automatically with that kind of parameters and add an extra "segment" to the breadcrumb.

Yes, but we'll also need to let people override title and breadcrumb, because for `/orders/:id` it would just match the '/orders' route and get an "Orders" title and "ACME > Orders" breadcrumb. While likely you want to Say something like "Order #123: 5 items"and "ACME > Orders > 123".

> Seems like a complicated problem, not sure if it would be too much to ask users to exclude "new" from the regex if they really want to do something like this... Or we could give priority to absolute paths? Just some first thoughts.

I think we can just match on the first route we find instead of errorring if we find two matching routes, like we do now. I'd potentially spec the algorithm as follows:

* we create a routing table which maps patterns to navigation items. Just an ordered Map of patterns=>navigation item
* we traverse the whole navigation. For every page item we calculate its path by joining its parent's segments with a slash. 
  * This is what this page item would link to when clicked
  * We push this path+item to the routing table
  * If the item contains a `pattern` property, we concatenate it with the parent path push this pattern+item combo to the routing table. We wont allow absolute paths for now (until user request) they can't start with a slash.
* Once fully built, we can obtain the current path from the router
* We iterate the routing table until we find the first pattern that matches the current path. 
  * This found item is the selected one in the navigation
  * Its title + breadcrumbs is used in pagecontainer
  * Any subsequent item that is matched is not considered as selected. We don't show an error. i.e. if We match `/orders/new` and there is also a `/orders/:id`, we only show `/orders/new` as selected


> I guess any regex could be used to match whatever people want? But we could accept a specific string too maybe I guess.

I wouldn't accept regexes (at least until there is a compelling use-case) and just use `path-to-regexp` instead.

**Comment by apedroferreira** [comment]
> I wouldn't accept regexes (at least until there is a compelling use-case) and just use path-to-regexp instead.

Yeah, I didn't remember immediately that that way of writing paths isn't exactly regex. We can start with that, it's a good standard.

About breadcrumbs, we can let people override the breadcrumb name per item, that sounds good. Maybe it will even help if somehow they can use a function there to get an actual readable segment name from the ID in the URL.

About matching routes, using the first match seems like a good start that we can expand on later if we find some more complicated scenarios. Will require some changes to the current logic but we can find an efficient way to do it for sure, your suggestion sounds good at first glance.

**Comment by jayenashar** [comment]
does #3991 work for parent navigation items?  i'm setting pattern to `''` and `'{/:segment}*'` but the item is neither highlighted nor initially expanded.  really i want it to be expanded by default on all routes, so if there is another way to achieve that, please let me know.

**Comment by apedroferreira** [comment]
> does [#3991](https://github.com/mui/toolpad/pull/3991) work for parent navigation items? i'm setting pattern to `''` and `'{/:segment}*'` but the item is neither highlighted nor initially expanded. really i want it to be expanded by default on all routes, so if there is another way to achieve that, please let me know.

Right now a navigation parent item should start as extended if an item inside it is selected/active.

Can you share your specific navigation? From what I've tested `pattern: ''` seems to kind of work like a wild card.

**Comment by jayenashar** [comment]
`Top 15` acts as a collection of quick submenu items, so i want it expanded when any page loads.  right now i have a `useLayoutEffect` that clicks on it.

```js
  const navigation = [
    {
      title: 'Dockets',
      segment: 'dockets',
      pattern: 'dockets{/:segment}*',
      action: (
        <Box sx={{ display: 'flex', gap: 1, alignItems: 'center' }}>
          {getIconsFromLoadingStatus(docketsLoadingStatus, styles)}
          <Chip label={dockets.length} color="primary" size="small" key="numDockets" />
        </Box>
      )
    },
    {
      title: 'Receivers',
      segment: 'receivers',
      pattern: 'receivers{/:segment}*',
      action: (
        <Box sx={{ display: 'flex', gap: 1, alignItems: 'center' }}>
          {getIconsFromLoadingStatus(receiversLoadingStatus, styles)}
          <Chip label={receivers.length} color="primary" size="small" key="numReceivers" />
        </Box>
      )
    },
    {
      kind: 'divider',
    },
    {
      title: 'Top 15',
      segment: 'dockets',
      pattern: '',
      children: top15Children.length > 0 ?
        top15Children.map(receiverData => ({
          title: `${receiverData.receiver}`,
          segment: receiverData.uncollectedDocket
            ? `${receiverData.uncollectedDocket.id}`
            : `${uuidv4()}${receiverData.receiverID ? `/receiverID/${receiverData.receiverID}` : `/receiver/${receiverData.receiver}`}`,
        })) : [
          {title: 'No recent dockets', segment: ''}
        ],
    },
  ];
```

**Comment by apedroferreira** [comment]
> `Top 15` acts as a collection of quick submenu items, so i want it expanded when any page loads. right now i have a `useLayoutEffect` that clicks on it.
> 
>   const navigation = [
>     {
>       title: 'Dockets',
>       segment: 'dockets',
>       pattern: 'dockets{/:segment}*',
>       action: (
>         <Box sx={{ display: 'flex', gap: 1, alignItems: 'center' }}>
>           {getIconsFromLoadingStatus(docketsLoadingStatus, styles)}
>           <Chip label={dockets.length} color="primary" size="small" key="numDockets" />
>         </Box>
>       )
>     },
>     {
>       title: 'Receivers',
>       segment: 'receivers',
>       pattern: 'receivers{/:segment}*',
>       action: (
>         <Box sx={{ display: 'flex', gap: 1, alignItems: 'center' }}>
>           {getIconsFromLoadingStatus(receiversLoadingStatus, styles)}
>           <Chip label={receivers.length} color="primary" size="small" key="numReceivers" />
>         </Box>
>       )
>     },
>     {
>       kind: 'divider',
>     },
>     {
>       title: 'Top 15',
>       segment: 'dockets',
>       pattern: '',
>       children: top15Children.length > 0 ?
>         top15Children.map(receiverData => ({
>           title: `${receiverData.receiver}`,
>           segment: receiverData.uncollectedDocket
>             ? `${receiverData.uncollectedDocket.id}`
>             : `${uuidv4()}${receiverData.receiverID ? `/receiverID/${receiverData.receiverID}` : `/receiver/${receiverData.receiver}`}`,
>         })) : [
>           {title: 'No recent dockets', segment: ''}
>         ],
>     },
>   ];


Thanks for sharing your navigation!
I see, it looks like if you set a `segment` in the parent item, even if you have a "catch-all" pattern in one of the children items it still needs to be in a path under the parent `segment` to show as selected, even if the parent also has a `pattern`.

So this seems to be a bug we can fix by making the `pattern` never depend on the parent items no matter how nested the navigation item is. I'll create an issue for it.

**Comment by apedroferreira** [comment]
I've created an issue for this: https://github.com/mui/toolpad/issues/4732
